File: clients/ruby/examples/basic_example.rb
#!/usr/bin/env ruby

require 'merklekv'

puts "MerkleKV Ruby Client Example"
puts "============================"
puts

begin
  # Create client with custom timeout
  client = MerkleKV::Client.new(host: "127.0.0.1", port: 7379, timeout: 5.0)

  puts "1. Basic Operations:"
  
  # Set a value
  client.set("user:1", "alice")
  puts "✓ Set user:1 = alice"

  # Get the value
  value = client.get("user:1")
  puts "✓ Get user:1 = #{value || '(nil)'}"

  # Delete the key
  deleted = client.delete("user:1")
  puts "✓ Delete user:1 = #{deleted}"

  # Try to get after delete
  after_delete = client.get("user:1")
  puts "✓ Get user:1 after delete = #{after_delete || '(nil)'}"

  puts
  puts "2. Special Values:"
  
  # Empty value (automatically handled as "")
  client.set("empty:test", "")
  empty_value = client.get("empty:test")
  puts "✓ Empty value: '#{empty_value}'"

  # Unicode value
  unicode_value = "🚀 Hello 世界! ñáéíóú"
  client.set("unicode:test", unicode_value)
  retrieved_unicode = client.get("unicode:test")
  puts "✓ Unicode value: #{retrieved_unicode}"

  # Value with spaces
  spaces_value = "value with multiple spaces"
  client.set("spaces:test", spaces_value)
  retrieved_spaces = client.get("spaces:test")
  puts "✓ Spaces value: '#{retrieved_spaces}'"

  # Clean up
  client.delete("empty:test")
  client.delete("unicode:test")
  client.delete("spaces:test")

  puts
  puts "3. Performance Test (1000 operations):"
  
  start_time = Time.now
  
  1000.times do |i|
    key = "perf:#{i}"
    value = "value#{i}"
    
    client.set(key, value)
    client.get(key)
    client.delete(key)
  end
  
  end_time = Time.now
  total_time = end_time - start_time
  
  avg_latency = (total_time * 1000) / 3000.0 # 3000 total operations
  puts "✓ Average latency: #{'%.2f' % avg_latency}ms per operation"
  
  if avg_latency < 5.0
    puts "✓ Performance target met (<5ms)"
  else
    puts "⚠️ Performance target not met (>5ms)"
  end

  puts
  puts "4. Connection Management:"
  
  # Show connection status
  puts "✓ Connected: #{client.connected?}"
  
  # Multiple operations on same connection
  5.times do |i|
    client.set("multi:#{i}", "value#{i}")
    value = client.get("multi:#{i}")
    client.delete("multi:#{i}")
    puts "✓ Operation #{i + 1} completed: #{value}"
  end

  # Close connection
  client.close
  puts "✓ Connection closed: #{client.connected?}"

rescue MerkleKV::ConnectionError => e
  puts "❌ Connection error: #{e.message}"
  puts "   Make sure MerkleKV server is running: cargo run --release"
  exit 1
rescue MerkleKV::TimeoutError => e
  puts "❌ Timeout error: #{e.message}"
  exit 1
rescue MerkleKV::ProtocolError => e
  puts "❌ Protocol error: #{e.message}"
  exit 1
rescue MerkleKV::Error => e
  puts "❌ MerkleKV error: #{e.message}"
  exit 1
rescue StandardError => e
  puts "❌ Unexpected error: #{e.message}"
  exit 1
ensure
  # Ensure client is closed
  client&.close
end

puts
puts "✅ Example completed successfully!"

File: clients/cpp/examples/basic_example.cpp
#include <iostream>
#include <iomanip>
#include <chrono>
#include <string>
#include <merklekv/merklekv.hpp>

using namespace merklekv;

int main() {
    std::cout << "MerkleKV C++ Client Example\n";
    std::cout << "===========================\n\n";

    try {
        // Create client with custom timeout
        MerkleKvClient client{"127.0.0.1", 7379, std::chrono::milliseconds(5000)};
        
        std::cout << "1. Basic Operations:\n";
        
        // Set a value
        client.set("user:1", "alice");
        std::cout << "✓ Set user:1 = alice\n";

        // Get the value
        auto value = client.get("user:1");
        std::cout << "✓ Get user:1 = " << (value ? *value : "(null)") << "\n";

        // Delete the key
        bool deleted = client.del("user:1");
        std::cout << "✓ Delete user:1 = " << (deleted ? "true" : "false") << "\n";

        // Try to get after delete
        auto afterDelete = client.get("user:1");
        std::cout << "✓ Get user:1 after delete = " << (afterDelete ? *afterDelete : "(null)") << "\n";

        std::cout << "\n2. Special Values:\n";
        
        // Empty value (automatically handled as "")
        client.set("empty:test", "");
        auto emptyValue = client.get("empty:test");
        std::cout << "✓ Empty value: '" << (emptyValue ? *emptyValue : "(null)") << "'\n";

        // Unicode value
        std::string unicodeValue = "🚀 Hello 世界! ñáéíóú";
        client.set("unicode:test", unicodeValue);
        auto retrievedUnicode = client.get("unicode:test");
        std::cout << "✓ Unicode value: " << (retrievedUnicode ? *retrievedUnicode : "(null)") << "\n";

        // Value with spaces
        std::string spacesValue = "value with multiple spaces";
        client.set("spaces:test", spacesValue);
        auto retrievedSpaces = client.get("spaces:test");
        std::cout << "✓ Spaces value: '" << (retrievedSpaces ? *retrievedSpaces : "(null)") << "'\n";

        // Clean up
        client.del("empty:test");
        client.del("unicode:test");
        client.del("spaces:test");

        std::cout << "\n3. Performance Test (1000 operations):\n";
        
        auto start = std::chrono::high_resolution_clock::now();
        
        for (int i = 0; i < 1000; ++i) {
            std::string key = "perf:" + std::to_string(i);
            std::string val = "value" + std::to_string(i);
            
            client.set(key, val);
            client.get(key);
            client.del(key);
        }
        
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
        
        double avgLatency = static_cast<double>(duration.count()) / 3000.0; // 3000 total operations
        std::cout << "✓ Average latency: " << std::fixed << std::setprecision(2) << avgLatency << "ms per operation\n";
        
        if (avgLatency < 5.0) {
            std::cout << "✓ Performance target met (<5ms)\n";
        } else {
            std::cout << "⚠️ Performance target not met (>5ms)\n";
        }

        std::cout << "\n4. Move Semantics:\n";
        
        // Demonstrate move semantics
        MerkleKvClient client1;
        client1.set("move:test", "movable value");
        
        MerkleKvClient client2 = std::move(client1);
        auto movedValue = client2.get("move:test");
        std::cout << "✓ Moved client value: " << (movedValue ? *movedValue : "(null)") << "\n";
        
        client2.del("move:test");

    } catch (const ConnectionException& e) {
        std::cerr << "❌ Connection error: " << e.what() << "\n";
        std::cerr << "   Make sure MerkleKV server is running: cargo run --release\n";
        return 1;
    } catch (const TimeoutException& e) {
        std::cerr << "❌ Timeout error: " << e.what() << "\n";
        return 1;
    } catch (const ProtocolException& e) {
        std::cerr << "❌ Protocol error: " << e.what() << "\n";
        return 1;
    } catch (const Exception& e) {
        std::cerr << "❌ MerkleKV error: " << e.what() << "\n";
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "❌ Standard error: " << e.what() << "\n";
        return 1;
    }

    std::cout << "\n✅ Example completed successfully!\n";
    return 0;
}
